modules/ximgproc/src/globalmatting.cpp

```diff
@@ -0,0 +1,568 @@
+// of this distribution and at http://opencv.org/license.html.
+#include "precomp.hpp"
+
+
+namespace cv
+{
+namespace ximgproc
+{
+template <typename T>
+inline T sqr(T a)
+{
+           return a * a;
+}
+
+struct IntensityComp
+{
+    IntensityComp(const cv::Mat_<cv::Vec3b> &img_temp) : img(img_temp)
+    {
+
+    }
+
+    bool operator()(const cv::Point &p0, const cv::Point &p1) const
+    {
+	const cv::Vec3b &c0 = img(p0.y, p0.x);
+	const cv::Vec3b &c1 = img(p1.y, p1.x);
+
+	return ((int)c0[0] + (int)c0[1] + (int)c0[2]) < ((int)c1[0] + (int)c1[1] + (int)c1[2]);
+    }
+
+    const cv::Mat_<cv::Vec3b> &img;
+};
+
+
+
+
+
+std::vector<cv::Point> GlobalMatting::findBoundaryPixels(const cv::Mat_<uchar> &trimap, int a, int b)
+{
+    std::vector<cv::Point> result;
+
+    for (int x = 1; x < trimap.cols - 1; ++x)
+	for (int y = 1; y < trimap.rows - 1; ++y)
+	{
+	    if (trimap(y, x) == a)
+	    {
+	        if (trimap(y - 1, x) == b ||
+	            trimap(y + 1, x) == b ||
+	            trimap(y, x - 1) == b ||
+	            trimap(y, x + 1) == b)
+	        {
+	            result.push_back(cv::Point(x, y));
+	        }
+	    }
+	}
+
+    return result;
+}
+
+// Eq. 2
+float GlobalMatting::calculateAlpha(const cv::Vec3b &F, const cv::Vec3b &B, const cv::Vec3b &I)
+{
+    float result = 0;
+    float div = 1e-6f;
+    for (int c = 0; c < 3; ++c)
+    {
+	float f = F[c];
+	float b = B[c];
+	float i = I[c];
+
+	result += (i - b) * (f - b);
+	div += (f - b) * (f - b);
+    }
+
+    return std::min(std::max(result / div, 0.f), 1.f);
+}
+
+// Eq. 3
+float GlobalMatting::colorCost(const cv::Vec3b &F, const cv::Vec3b &B, const cv::Vec3b &I, float alpha)
+{
+    float result = 0;
+    for (int c = 0; c < 3; ++c)
+    {
+	float f = F[c];
+	float b = B[c];
+	float i = I[c];
+
+	result += sqr(i - (alpha * f + (1 - alpha) * b));
+    }
+
+    return sqrt(result);
+}
+
+// Eq. 4
+float GlobalMatting::distCost(const cv::Point &p0, const cv::Point &p1, float minDist)
+{
+    int dist = sqr(p0.x - p1.x) + sqr(p0.y - p1.y);
+    return sqrt((float)dist) / minDist;
+}
+
+float GlobalMatting::colorDist(const cv::Vec3b &I0, const cv::Vec3b &I1)
+{
+    int result = 0;
+
+    for (int c = 0; c < 3; ++c)
+	result += sqr((int)I0[c] - (int)I1[c]);
+
+    return sqrt((float)result);
+}
+
+float GlobalMatting::nearestDistance(const std::vector<cv::Point> &boundary, const cv::Point &p)
+{
+    int minDist2 = INT_MAX;
+    for (std::size_t i = 0; i < boundary.size(); ++i)
+    {
+	int dist2 = sqr(boundary[i].x - p.x)  + sqr(boundary[i].y - p.y);
+	minDist2 = std::min(minDist2, dist2);
+    }
+
+    return sqrt((float)minDist2);
+}
+
+
+
+
+void GlobalMatting::expansionOfKnownRegions(const cv::Mat_<cv::Vec3b> &image,
+	                            cv::Mat_<uchar> &trimap,
+	                            int r, float c)
+{
+    int w = image.cols;
+    int h = image.rows;
+
+    for (int x = 0; x < w; ++x)
+	for (int y = 0; y < h; ++y)
```

Comment:
"(x,y)" for loops have slowest memory access pattern. Any reason for that?

Reply:
Could you suggest an alternative ?

Reply:
```
for (int y = 0; y < h; ++y)
    for (int x = 0; x < w; ++x)
```

---

Don't use "tabs" in code